cql3/restrictions/statement_restrictions.cc

```diff
@@ -969,5 +975,229 @@ bool single_column_restriction::LIKE::is_satisfied_by(bytes_view data, const que
     return (*_matcher)(data);
 }
 
+namespace wip {
+
+namespace {
+
+using cql3::selection::selection;
+
+/// Returns cdef's value from the fetched data.
+bytes_opt get_value(const column_definition* cdef, const selection& selection,
+                    const std::vector<bytes>& partition_key, const std::vector<bytes>& clustering_key,
+                    const std::vector<bytes_opt>& other_columns) {
+    switch (cdef->kind) {
+    case column_kind::partition_key:
+        return partition_key[cdef->id];
+    case column_kind::clustering_key:
+        return clustering_key[cdef->id];
+    case column_kind::static_column:
+    case column_kind::regular_column:
+        return other_columns[selection.index_of(*cdef)];
+    default:
+        throw exceptions::unsupported_operation_exception("Unknown column kind");
+    }
 }
+
+/// True iff columns' values equal t.
+bool equal(::shared_ptr<term> t, const idents& columns, const selection& selection,
+           const std::vector<bytes>& partition_key, const std::vector<bytes>& clustering_key,
+           const std::vector<bytes_opt>& other_columns, const query_options& options) {
+    if (auto multi = dynamic_pointer_cast<multi_item_terminal>(t)) {
+        const auto& rhs = multi->get_elements();
+        if (rhs.size() != columns.size()) {
+            throw std::logic_error("LHS and RHS size mismatch");
+        }
+        for (size_t i = 0; i < rhs.size(); ++i) {
+            if (rhs[i] != get_value(columns[i], selection, partition_key, clustering_key, other_columns)) {
```

Comment:
I don't know if `!=` is sufficient. Consider rhs[i] = -0.0 and get_value() = +0.0. They compare equal using the type comparator and not equal using bitwise comparison.

Of course we should follow existing code practice first, even if it's wrong.

Reply:
> I don't know if `!=` is sufficient. Consider rhs[i] = -0.0 and get_value() = +0.0. They compare equal using the type comparator and not equal using bitwise comparison.

Turns out they don't, see below...

> Of course we should follow existing code practice first, even if it's wrong.

I don't know if this is intentional, but it looks like only UDTs don't use bitwise equality:
[types.cc:766](https://github.com/scylladb/scylla/blob/master/types.cc#L766)

Reply:
I don't understand. All types return false except for a select few. In particular, floating types return false.

@espindola I guess we can expand it for map and list when both keys and values are byte-order equal.

Reply:
Oh, floating types are simple types. So that's a bug @espindola. In any case, whether to compare bytewise or using the type comparator is a decision for the type. not for a rewrite of CQL3 restriction representation.

Reply:
Right, I'll change this to use the type comparator.  I just don't have a good way to test it until that bug is fixed.

Reply:
Switched to type comparator.  Will file a bug for -0.0=+0.0.